docs/client-sdk-states-and-lifecycle.mdx

```diff
@@ -0,0 +1,414 @@
+
+Nimbus will also emit Glean [events](https://mozilla.github.io/glean/book/user/metrics/event.html) when the client code calls experiment-related APIs:
+
+
+
+*   An "**_exposure_**" event, when client code calls `activateExperiment` or `isFeatureEnabled` for the first time, and the experiment is in **Enrolled** state.
+    *   Event field "experiment" records the experiment slug.
+    *   Event field "branch" records the branch into which the client enrolled.
+    *   Event field "enrollmentId" contains a randomly-generated identifier for this enrollment.
+
+
+## Client Behaviours
+
+
+#### When asked if an experiment is active
+
+
+
+*   Get the current client state and last-seen server state for the named experiment.
+*   If the client state is **Enrolled**, return _true_.
+*   Otherwise, return _false_.
+
+
+#### When asked for the current branch of an experiment
+
+
+
+*   Get the current client state and last-seen server state for the named experiment.
+*   If the client state is **Enrolled**, return the enrolled branch and its config.
+*   Otherwise, return _None_.
+
+
+#### When asked for the current config of a feature
+
+
+
+*   Get the current client state and last-seen server state for all experiments.
+*   Discard any experiments whose client state is not **Enrolled**.
+*   For each remaining experiment:
+    *   Get the branch config for the **Enrolled **branch.
+    *   If it does not contain config for the named feature, discard the experiment
+*   If more than one experiment remains, TODO this is an error right? report it somehow.
+*   If exactly one experiment remains, return the feature config from its **Enrolled** branch.
+*   Otherwise, return _None_.
+
+
+#### When a server update transitions an experiment to "Enrolling"
+
+
+
+*   If the current client state for the experiment is **Unknown** or **NotEnrolled**:
+    *   Get the current state for all known experiments that conflict with this one.
+    *   Evaluate targeting/bucketing/etc, if it says to enroll, set client state to **Enrolled**.
+        *   TODO: trigger a notification in the client app somehow, to redraw UI etc?
+    *   Otherwise, set client state to **NotEnrolled**.
+*   If the current client state for the experiment is **Enrolled**:
+    *   TODO: what exactly? How does an existing enrollment change in response to updated experiment config?
+*   Otherwise, set the current client state for the experiment to **Errored** and log some diagnostic telemetry.
+
+
+#### When a server update transitions an experiment to "Paused?" or "Running"
+
+
+
+*   If the current client state for the experiment is **Enrolled**:
+    *   TODO: trigger a notification in the client app somehow, to redraw UI etc?
+
+
+#### When a server update transitions an experiment to "Cancelled" or "Completed"
+
+
+
+*   If the current client state for the experiment is **Enrolled**:
+    *   TODO: trigger a notification in the client app somehow, to redraw UI etc?
```

Comment:
Same comment as above, the app controls the application of experiment enrollment and should handle it accordingly.  The notification of the app seems unnecessary to me.

Reply:
Now tracked in SDK-303